File: backend/src/main/java/br/ufpr/sept/so2/modules/iam/application/LogoutUseCase.java
package br.ufpr.sept.so2.modules.iam.application;

import br.ufpr.sept.so2.modules.iam.application.ports.OpaqueTokenHasher;
import br.ufpr.sept.so2.modules.iam.application.ports.RefreshTokenRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.OffsetDateTime;

@Service
public class LogoutUseCase {

    private final RefreshTokenRepository refreshTokenRepository;
    private final OpaqueTokenHasher opaqueTokenHasher;

    public LogoutUseCase(
            RefreshTokenRepository refreshTokenRepository,
            OpaqueTokenHasher opaqueTokenHasher
    ) {
        this.refreshTokenRepository = refreshTokenRepository;
        this.opaqueTokenHasher = opaqueTokenHasher;
    }

    @Transactional
    public void execute(String refreshRaw) {
        if (refreshRaw == null || refreshRaw.isBlank()) {
            return;
        }
        refreshTokenRepository.lockByTokenHash(opaqueTokenHasher.hash(refreshRaw))
                .ifPresent(sessao -> {
                    sessao.revogar(OffsetDateTime.now());
                    refreshTokenRepository.save(sessao);
                });
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/iam/application/LogoutUseCase.kt
package br.ufpr.sept.so2.modules.iam.application

import br.ufpr.sept.so2.modules.iam.application.ports.OpaqueTokenHasher
import br.ufpr.sept.so2.modules.iam.application.ports.RefreshTokenRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.OffsetDateTime

@Service
class LogoutUseCase(
    private val refreshTokenRepository: RefreshTokenRepository,
    private val opaqueTokenHasher: OpaqueTokenHasher,
) {
    @Transactional
    fun execute(refreshRaw: String?) {
        if (refreshRaw.isNullOrBlank()) {
            return
        }
        refreshTokenRepository.lockByTokenHash(opaqueTokenHasher.hash(refreshRaw))
            .ifPresent { sessao ->
                sessao.revogar(OffsetDateTime.now())
                refreshTokenRepository.save(sessao)
            }
    }
}
